pkg/binding/spec/attributes.go

```diff
@@ -0,0 +1,110 @@
+package spec
+
+import (
+	"errors"
+	"time"
+
+	ce "github.com/cloudevents/sdk-go/pkg/cloudevents"
+)
+
+// Accessor is a version-independent identifier and accessor for a cloudevents attribute.
+// Accessor values are comparable and can be used to index an array.
+type Accessor uint8
+
+func (i Accessor) Get(c ce.EventContextReader) interface{}          { return getter[i](c) }
+func (i Accessor) Set(c ce.EventContextWriter, v interface{}) error { return setter[i](c, v) }
+func (i Accessor) String() string                                   { return defaultName[i] }
+func (i Accessor) IsRequired() bool                                 { return i < DataContentType }
+
+// Accessor constants.
+const (
+	// Required attributes
+	ID Accessor = iota
+	Source
+	SpecVersion
+	Type
+	// Optional attributes
+	DataContentType
+	DataSchema
+	Subject
+	Time
+)
+const (
+	nAttrs = int(Time) + 1
+)
+
+// ErrBadType returned when there is an attribute/value type mismatch.
+var ErrBadType = errors.New("bad context attribute type")
+
+// Attribute is a named attribute accessor, the attribute name is specific to a Version.
+type Attribute struct {
+	Name string
+	Accessor
+}
+
+func (a Attribute) String() string { return a.Name }
+
+// Shortcuts, these are used a lot below.
+type ecR = ce.EventContextReader
+type ecW = ce.EventContextWriter
+
+var (
+	getter = [nAttrs]func(ecR) interface{}{
```

Comment:
what is this?

Reply:
it's an array of get(EventContextReader) interface{}  functions that wrap each EventContext.Get...() method so you can look up a getter by index (which can be looked up from a version-specific string name) You could do the same with reflection but it'd be more long winded and less efficient. I may have overdone it on the abbreviations, I'll try to make it more obvious :)

Reply:
Cleaned up on latest push: collect get/set in structs rather than separate function tables. Neater and less mysterious